### PR TITLE
feat(src/index): support synchronous config loading (`rc.sync`)

### LINKED ISSUE #184: Update cosmiconfig to latest version
Cosmiconfig below `5.0.0` has` js-yaml` deps package that has vulnerability issue. So please update it that postcss-load-config has no vulnerabilitiy too.

### PR DESCRIPTION
**Pull request by bmatcuk**
### `Notable Changes`

This PR adds support for a synchronous loading function, such as:

```js
const postcssrc = require("postcss-load-config")
const config = postcssrc.sync({})
```

I'm trying to write an eslint plugin, but eslint has no support for asynchronous plugins. I would love to use postcss-load-config, but I would need this support for that.

While I was in there, I also updated cosmiconfig to the latest version, which closes #184.

#### `Commit Message Summary (CHANGELOG)`

```
chore(package): update `cosmiconfig` v4.0.0...5.0.0
feat(src/index): support synchronous config loading (`rc.sync`)
```

### `Type`

- [x] Chore
- [x] Feature

### `SemVer`

- [x] Feature (:label: Minor)

### `Issues`

- Fixes #184 

### `Checklist`

- [x] Lint and unit tests pass with my changes
- [x] I have added tests that prove my fix is effective/works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes are merged and published in downstream modules

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -29,51 +54,64 @@ const rc = (ctx, path, options) => {
...
+  path = path ? resolve(path) : process.cwd()
 
-      return {
-        plugins: loadPlugins(config, file),
-        options: loadOptions(config, file),
-        file: file
-      }
-    })
+  const result = config('postcss', options).searchSync(path)
+  if (!result) {
```
Comment (michael-ciniawsky) [inline src/index.js]:
```suggestion

  if (!result) {
```

File: src/options.js [hunk 2]
```diff
@@ -2,6 +2,8 @@
 
 const req = require('import-cwd')
 
+const { stripRequireStack } = require('./utils.js')
```
Comment (michael-ciniawsky) [inline src/options.js]: Why is this needed ?  🙃

File: src/plugins.js [hunk 3]
```diff
@@ -14,22 +16,18 @@ const req = require('import-cwd')
  * @return {Function} PostCSS Plugin
  */
 const load = (plugin, options, file) => {
-  if (
-    options === null ||
-    options === undefined ||
-    Object.keys(options).length === 0
-  ) {
-    try {
+  try {
```
Comment (michael-ciniawsky) [inline src/plugins.js]: 👍

File: test/Errors.test.js [hunk 4]
```diff
@@ -52,6 +62,58 @@ describe('Loading Plugins - {Error}', () => {
   })
 })
 
+describe('Loading Plugins Sync - {Error}', () => {
```
Comment (michael-ciniawsky) [inline test/Errors.test.js]:
```suggestion
describe('Loading Plugins - Sync - {Error}', () => {
```

File: test/Errors.test.js [hunk 5]
```diff
@@ -10,6 +10,16 @@ test('Loading Config - {Error}', () => {
   })
 })
 
+test('Loading Config Sync - {Error}', () => {
```
Comment (michael-ciniawsky) [inline test/Errors.test.js]:
```suggestion
test('Loading Config - Sync - {Error}', () => {
```

File: test/js.test.js [hunk 6]
```diff
@@ -60,13 +69,13 @@ test('postcss.config.js - {Object} - Process SSS', () => {
...
 
-test('postcss.config.js - {Array} - Load Config', () => {
+describe('postcss.config.js - {Array} - Load Config', () => {
   const ctx = {
     parser: true,
     syntax: true
   }
 
-  return postcssrc(ctx, 'test/js/array').then((config) => {
+  const testConfig = (config) => {
```
Comment (michael-ciniawsky) [inline test/js.test.js]:
```suggestion
  const expected = (config) => {
```

File: test/pkg.test.js [hunk 7]
```diff
@@ -7,8 +7,8 @@ const postcssrc = require('../src/index.js')
 
 const { fixture, expected } = require('./utils.js')
 
-test('package.json - {Object} - Load Config', () => {
-  return postcssrc({}, 'test/pkg').then((config) => {
+describe('package.json - {Object} - Load Config', () => {
+  const testConfig = (config) => {
```
Comment (michael-ciniawsky) [inline test/pkg.test.js]:
```suggestion
  const expected = (config) => {
```

File: test/rc.test.js [hunk 8]
```diff
@@ -7,8 +7,8 @@ const postcssrc = require('../src/index.js')
 
 const { fixture, expected } = require('./utils.js')
 
-test('.postcssrc - {Object} - Load Config', () => {
-  return postcssrc({}, 'test/rc').then((config) => {
+describe('.postcssrc - {Object} - Load Config', () => {
+  const testConfig = (config) => {
```
Comment (michael-ciniawsky) [inline test/rc.test.js]:
```suggestion
  const expected = (config) => {
```

File: test/utils.test.js [hunk 9]
```diff
@@ -0,0 +1,16 @@
+'use strict'
```
Comment (michael-ciniawsky) [inline test/utils.test.js]: Please remove this file, you don't need to add a test for a testing util :)

File: test/js.test.js [hunk 10]
```diff
@@ -7,13 +7,13 @@ const postcssrc = require('../src/index.js')
...
 
-test('postcss.config.js - {Object} - Load Config', () => {
+describe('postcss.config.js - {Object} - Load Config', () => {
   const ctx = {
     parser: true,
     syntax: true
   }
 
-  return postcssrc(ctx, 'test/js/object').then((config) => {
+  const testConfig = (config) => {
```
Comment (michael-ciniawsky) [inline test/js.test.js]:
```suggestion
  const expected = (config) => {
```

File: test/Errors.test.js [hunk 11]
```diff
@@ -77,3 +139,35 @@ describe('Loading Options - {Error}', () => {
     })
   })
 })
+
+describe('Loading Options Sync - {Error}', () => {
```
Comment (michael-ciniawsky) [inline test/Errors.test.js]:
```suggestion
describe('Loading Options - Sync - {Error}', () => {
```

File: test/pkg.test.js [hunk 12]
```diff
@@ -21,6 +21,15 @@ test('package.json - {Object} - Load Config', () => {
...
       .toEqual(path.resolve('test/pkg', 'package.json'))
+  }
+
+  test('Async', () => {
+    return postcssrc({}, 'test/pkg').then(testConfig)
+  })
+
+  test('Sync', () => {
+    const config = postcssrc.sync({}, 'test/pkg')
+    testConfig(config)
```
Comment (michael-ciniawsky) [inline test/pkg.test.js]:
```suggestion

    expected(config)
```

File: test/rc.test.js [hunk 13]
```diff
@@ -21,6 +21,15 @@ test('.postcssrc - {Object} - Load Config', () => {
...
       .toEqual(path.resolve('test/rc', '.postcssrc'))
+  }
+
+  test('Async', () => {
+    return postcssrc({}, 'test/rc').then(testConfig)
+  })
+
+  test('Sync', () => {
+    const config = postcssrc.sync({}, 'test/rc')
+    testConfig(config)
```
Comment (michael-ciniawsky) [inline test/rc.test.js]:
```suggestion
    
    expected(config)
```

File: test/js.test.js [hunk 14]
```diff
@@ -79,6 +88,15 @@ test('postcss.config.js - {Array} - Load Config', () => {
...
       .toEqual(path.resolve('test/js/array', 'postcss.config.js'))
+  }
+
+  test('Async', () => {
+    return postcssrc(ctx, 'test/js/array').then(testConfig)
+  })
+
+  test('Sync', () => {
+    const config = postcssrc.sync(ctx, 'test/js/array')
+    testConfig(config)
```
Comment (michael-ciniawsky) [inline test/js.test.js]:
```suggestion
    
    expected(config)
```

File: test/js.test.js [hunk 15]
```diff
@@ -26,6 +26,15 @@ test('postcss.config.js - {Object} - Load Config', () => {
...
       .toEqual(path.resolve('test/js/object', 'postcss.config.js'))
+  }
+
+  test('Async', () => {
+    return postcssrc(ctx, 'test/js/object').then(testConfig)
+  })
+
+  test('Sync', () => {
+    const config = postcssrc.sync(ctx, 'test/js/object')
+    testConfig(config)
```
Comment (michael-ciniawsky) [inline test/js.test.js]:
```suggestion
    
    expected(config)
```

**Comment by michael-ciniawsky** [comment]
Please split the commit into to separate commits

```
chore(package): update `cosmiconfig` v4.0.0...5.0.0
feat(src/index): support synchronous config loading (`rc.sync`)
```

File: src/options.js (hunk 2 above)
Comment (bmatcuk) [reply src/options.js]: Tests were failing because when require() failed, it was throwing an error that looked something like:

```
Failed to load module: 'whatever'
Require stack:
- ...
- ...
- ...
```

It seems the original code was not expecting that "require stack" stuff, so the tests were failing. I'm not sure if it's because I'm using a newer version of node or what, but I wrote a utility function to strip that stuff out. It's not really useful in the error message anyway, since the stack is deep in postcss-load-config.

Comment (michael-ciniawsky) [reply src/options.js]: ahhh ok so 'require stack' refers to the Error Stacktrace (`err.stack`) in general? Never saw the `Require stack: ...` prefix before tbh, maybe cosmiconfig formats the `{Error}` message 🤷‍♂. Rename it to `stripStack` or maybe even `format` please

File: src/utils.js [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+ * Strips "Require stack" from error messages
+ *
+ * @private
+ * @method stripRequireStack
+ *
+ * @param  {Error} err Error object
+ *
+ * @return {String} The error message with any "require stack" striped out
+ */
+const stripRequireStack = (err) => {
```
Comment (michael-ciniawsky) [inline src/utils.js]:
```js
const stripStack = (err) => {
  return err.message.split('\n')[0]
}
```

File: src/options.js (hunk 2 above)
Comment (michael-ciniawsky) [reply src/options.js]: 😨 Wait a sec... I just saw this modifies the `{Error}` itself and doesn't strip off the satck for tests only 🤦‍♂. We need to be careful here as this might introduce a breaking change. Which test where failing in particular?

Personally agreeing with the `stack` being rather uninteresting here.

**src/Error.js**
```js
class ConfigError extends Error {
   constructor (message) {
     super()
     
     this.name = 'ConfigError'
     this.message = message
     // this.stack = false
       
     Error.captureStackTrace(this, this.constructor)
   }
}

module.exports = ConfigError
```

Comment (bmatcuk) [reply src/options.js]: The "Require stack" appears in the error message. It's not actually modifying the error, it's just taking the message from the error and stripping out the "require stack" stuff.

Pretty much every test in Errors.test.js was failing without this change, though, it's weird... I just tried it on a different laptop to get you some example test output and it worked fine, so...

¯\\\_(ツ)\_/¯

I'll revert this part of my PR since I can no longer reproduce the issue...

**Comment by bmatcuk** [comment]
Made all of the requested changes; removed the "strip require stack" stuff since I can't reproduce it on another laptop

**Comment by bmatcuk** [comment]
lol, travis is failing with the require stack error now! What do you guys think would be the best way to handle it?

File: test/Errors.test.js [hunk 17]
```diff
@@ -77,3 +139,35 @@ describe('Loading Options - {Error}', () => {
...
   })
 })
+
+describe('Loading Options - Sync - {Error}', () => {
+  test('Parser - {String}', () => {
+    try {
+      postcssrc.sync({}, 'test/err/options/parser')
+    } catch (err) {
+      expect(err.message).toMatch(
+        /^Loading PostCSS Parser failed: (.*)\n\n\(@.*\)$/
```
Comment (michael-ciniawsky) [inline test/Errors.test.js]:
```suggestion
        /^Loading PostCSS Parser failed: (.*|\n*)/
```

File: test/Errors.test.js [hunk 18]
```diff
@@ -77,3 +139,35 @@ describe('Loading Options - {Error}', () => {
...
+      )
+    }
+  })
+
+  test('Syntax - {String}', () => {
+    try {
+      postcssrc.sync({}, 'test/err/options/syntax')
+    } catch (err) {
+      expect(err.message).toMatch(
+        /^Loading PostCSS Syntax failed: (.*)\n\n\(@.*\)$/
```
Comment (michael-ciniawsky) [inline test/Errors.test.js]:
```suggestion
        /^Loading PostCSS Syntax failed: (.*|\n*)/
```

File: test/Errors.test.js [hunk 19]
```diff
@@ -77,3 +139,35 @@ describe('Loading Options - {Error}', () => {
...
+      )
+    }
+  })
+
+  test('Stringifier - {String}', () => {
+    try {
+      postcssrc.sync({}, 'test/err/options/stringifier')
+    } catch (err) {
+      expect(err.message).toMatch(
+        /^Loading PostCSS Stringifier failed: (.*)\n\n\(@.*\)$/
```
Comment (michael-ciniawsky) [inline test/Errors.test.js]:
```suggestion
        /^Loading PostCSS Stringifier failed: (.*|\n*)/
```

File: test/Errors.test.js [hunk 20]
```diff
@@ -52,6 +62,58 @@ describe('Loading Plugins - {Error}', () => {
...
+      )
+    }
+  })
+
+  test('Plugin - {Object} - Options', () => {
+    try {
+      postcssrc.sync({}, 'test/err/plugins/object/options')
+    } catch (err) {
+      expect(err.message).toMatch(
+        /^Loading PostCSS Plugin failed: (.*)\n\n\(@.*\)$/
```
Comment (michael-ciniawsky) [inline test/Errors.test.js]:
```suggestion
        /^Loading PostCSS Plugin failed: (.*|\n*)/
```

File: test/Errors.test.js [hunk 21]
```diff
@@ -52,6 +62,58 @@ describe('Loading Plugins - {Error}', () => {
...
+      )
+    }
+  })
+
+  test('Plugin - {Object}', () => {
+    try {
+      postcssrc.sync({}, 'test/err/plugins/object')
+    } catch (err) {
+      expect(err.message).toMatch(
+        /^Loading PostCSS Plugin failed: (.*)\n\n\(@.*\)$/
```
Comment (michael-ciniawsky) [inline test/Errors.test.js]:
```suggestion
        /^Loading PostCSS Plugin failed: (.*|\n*)/
```

File: test/Errors.test.js [hunk 22]
```diff
@@ -52,6 +62,58 @@ describe('Loading Plugins - {Error}', () => {
   })
 })
 
+describe('Loading Plugins - Sync - {Error}', () => {
+  test('Plugin - {Type} - Invalid', () => {
+    try {
+      postcssrc.sync({}, 'test/err/plugins')
+    } catch (err) {
+      expect(err.message).toMatch(
+        /^Invalid PostCSS Plugin found at: (.*)\n\n\(@.*\)$/
```
Comment (michael-ciniawsky) [inline test/Errors.test.js]:
```suggestion
        /^Invalid PostCSS Plugin found at: (.*|\n*)/
```

**Comment by bmatcuk** [comment]
Thanks @michael-ciniawsky - I've updated the tests (I went with `/^blah blah .*$/m` instead). Looks like all the tests are passing now =)

File: test/Errors.test.js (hunk 22 above)
Comment (bmatcuk) [reply test/Errors.test.js]: This one doesn't need updating =)

**Comment by michael-ciniawsky** [comment]
Released in [`v2.1.0`](https://github.com/michael-ciniawsky/postcss-load-config/releases/tag/v2.1.0) 🎉

**Comment by bmatcuk** [comment]
Thanks!
